File: src/problemset/divideAndConquer/MajorityElement.java
package problemset.divideAndConquer;

/*
 * @date_time: 2020-04-30 14:39
 * @author: MoeexT
 * @problem: 面试题39
 * @url: https://leetcode-cn.com/problems/shu-zu-zhong-chu-xian-ci-shu-chao-guo-yi-ban-de-shu-zi-lcof/
 */

public class MajorityElement {
    public int majorityElement(int[] nums) {
        int ans = 0;
        Integer former = null;

        for (int num : nums) {
            if (ans == 0) {
                former = num;
            }
            ans += (num == former) ? 1 : -1;
        }
        return former;
    }

    public static void main(String[] args) {
        MajorityElement solution = new MajorityElement();
        System.out.println(solution.majorityElement(new int[]{1,2,3,2,2,2,5,4,2}));
    }
}

File: src/problemset/array/MajorityElement.py
# https://leetcode-cn.com/problems/majority-element/
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        """
        84ms: 77.80%
        15.1MB: 5.04%
        """
        dic = {}
        for num in nums:
            dic.update({num: dic.get(num, 0) + 1})
        
        for key, value in dic.items():
            if value > len(nums) >> 1:
                return key


if __name__ == "__main__":
    sol = Solution()
    print(sol.majorityElement([]))
